src/tree/binary_tree/mod.rs

```diff
@@ -0,0 +1,310 @@
+#![allow(dead_code)]
+use std::{cmp::Ordering, convert::TryFrom};
+
+use crate::{tree::index::LeafIndex, tree::index::NodeIndex};
+
+pub(crate) use serde::{Deserialize, Serialize};
+
+use self::errors::BinaryTreeError;
+use self::utils::*;
+
+pub(crate) mod errors;
+#[cfg(test)]
+pub(crate) mod test_binary_tree;
+pub(crate) mod utils;
+
+/// A full binary tree in the array (vector) representation used in the MLS
+/// spec. Note, that this is not a full implementation of a binary tree, but
+/// rather only enables the operations needed by MLS.
+#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct BinaryTree<T: PartialEq> {
+    nodes: Vec<T>,
+}
+
+impl<T: PartialEq> TryFrom<Vec<T>> for BinaryTree<T> {
+    type Error = BinaryTreeError;
+
+    /// Create a binary tree from a vector of nodes. Throws an error if the
+    /// number of given nodes can not be represented as a full tree.
+    fn try_from(value: Vec<T>) -> Result<Self, Self::Error> {
+        if value.len() % 3 != 0 && value.len() != 1 {
+            return Err(BinaryTreeError::TreeNotFull);
+        }
+        Ok(BinaryTree { nodes: value })
+    }
+}
+
+impl<T: PartialEq> BinaryTree<T> {
+    /// Get the size of the tree.
+    pub(crate) fn size(&self) -> NodeIndex {
+        NodeIndex::from(self.nodes.len())
+    }
+
+    /// Get the number of leaves in the tree.
+    pub(crate) fn leaf_count(&self) -> LeafIndex {
+        LeafIndex::from((self.size().as_usize() + 1) / 2)
+    }
+
+    /// Get the index corresponding to the root of the tree.
+    pub(crate) fn root(&self) -> NodeIndex {
```

Comment:
I'm not sure I understand the rational behind this code. You have a tree with nodes but all your functions only return indices. So the actual tree content isn't used and users have to get and index first before they can access something? This is basically treemath put in a struct.

Reply:
Good point. It's not straight-forward to find a good interface to the tree, though, as in a lot of places, MLS relies on indices to determine which nodes to access in a tree, so we can't hide them completely. How should the layer above the BinaryTree indicate which node to operate on? We could use public keys, but unless something changes in the spec, we will always have to expose indices as well.